ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BatchDeleteHandler.java

```diff
@@ -0,0 +1,189 @@
+      RestRequest restRequest, DeleteBlobHandler deleteBlobHandler, RestResponseChannel restResponseChannel,
+      Callback<ReadableStreamChannel> callback) {
+
+    return buildCallback(frontendMetrics.deleteBlobRouterMetrics, bytesRead -> {
+      if (bytesRead == null) {
+        // Handle the case where bytesRead is null
+        throw new IllegalStateException("bytesRead is null");
+      }
+      // Deserialize the request body into a S3BatchDeleteObjects
+      S3MessagePayload.S3BatchDeleteObjects deleteRequest = deserializeRequest(channel);
+
+      if (deleteRequest != null) {
+        // Extract request path
+        RequestPath requestPath = (RequestPath) restRequest.getArgs().get(RestUtils.InternalKeys.REQUEST_PATH);
+        Map<String, CompletableFuture<Void>> deleteFutures = new HashMap<>();
+
+        // Process each delete key in the batch
+        for (S3MessagePayload.S3BatchDeleteKeys object : deleteRequest.getObjects()) {
+
+          // Construct the delete path for each object
+          String singleDeletePath = requestPath.getPathAfterPrefixes() + "/" + object.getKey();
+
+          // Create a new RequestPath for the delete operation
+          List<String> emptyList = new ArrayList<>();
+          RequestPath newRequestPath =
+              RequestPath.parse(singleDeletePath, restRequest.getArgs(), emptyList, requestPath.getClusterName());
+          WrappedRestRequest singleDeleteRequest = new WrappedRestRequest(restRequest);
+          singleDeleteRequest.setArg(RestUtils.InternalKeys.REQUEST_PATH, newRequestPath);
+
+
+          NoOpResponseChannel noOpResponseChannel = new NoOpResponseChannel();
+          CompletableFuture<Void> future = new CompletableFuture<>();
+
+          // Handle the delete operation using the deleteBlobHandler
+          deleteBlobHandler.handle(singleDeleteRequest, noOpResponseChannel, new Callback<Void>() {
+            @Override
+            public void onCompletion(Void result, Exception exception) {
+              // Call our custom onDeleteCompletion to track success/failure
+              Map<String, Object> args = restRequest.getArgs();
+              restRequest.getArgs().remove("ambry-internal-key-target-account");
+              restRequest.getArgs().remove("ambry-internal-key-keep-alive-on-error-hint");
+              restRequest.getArgs().remove("ambry-internal-key-target-container");
+              boolean success = exception == null;
+              onDeleteCompletion(success, object.getKey());
+              future.complete(null);
+            }
+          });
+          deleteFutures.put(object.getKey(), future);
+        }
+
+        // Wait for all delete operations to complete
+        CompletableFuture<Void> allDeletes = CompletableFuture.allOf(deleteFutures.values().toArray(new CompletableFuture[0]));
+        try {
+          // Block until all futures complete
+          allDeletes.get();  // This will throw an exception if any future fails
```

Comment:
We don't want to block here, get() would block.
We want to do something like 
```
CompletableFuture.allOf(deletedFutures.values().toArray(new CompletableFuture[0])).whenComplete((v, exception) -> {
    // we should ignore exception, since we catch errors in the deleted and errors fields.
    // final callback
    if (failedRequest) {
        ....
    } else {
        // construct response
        callback.onComplete(xxx);
    }
});
```

In this way, the completable future is not going to block.